File: src/main/java/org/example/eshop/service/CorrelationIdService.java
package org.example.eshop.service;

import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class CorrelationIdService {
    private static final ThreadLocal<String> CORRELATION_ID = new ThreadLocal<>();
    public static final String CORRELATION_ID_HEADER = "X-Correlation-ID";

    public String generateCorrelationId() {
        return UUID.randomUUID().toString();
    }

    public void setCorrelationId(String correlationId) {
        CORRELATION_ID.set(correlationId);
    }

    public String getCorrelationId() {
        return CORRELATION_ID.get();
    }

    public void clearCorrelationId() {
        CORRELATION_ID.remove();
    }

    public String getOrGenerateCorrelationId() {
        String current = getCorrelationId();
        if (current != null) return current;
        String generated = generateCorrelationId();
        setCorrelationId(generated);
        return generated;
    }
}

File: src/main/kotlin/org/example/eshop/service/CorrelationIdService.kt
package org.example.eshop.service

import org.springframework.stereotype.Service
import java.util.*

@Service
class CorrelationIdService {
    
    companion object {
        private val correlationIdThreadLocal = ThreadLocal<String>()
        const val CORRELATION_ID_HEADER = "X-Correlation-ID"
    }
    
    fun generateCorrelationId(): String {
        return UUID.randomUUID().toString()
    }
    
    fun setCorrelationId(correlationId: String) {
        correlationIdThreadLocal.set(correlationId)
    }
    
    fun getCorrelationId(): String? {
        return correlationIdThreadLocal.get()
    }
    
    fun clearCorrelationId() {
        correlationIdThreadLocal.remove()
    }
    
    fun getOrGenerateCorrelationId(): String {
        return getCorrelationId() ?: generateCorrelationId().also { setCorrelationId(it) }
    }
}